Mjml.Net/Includes/Filesystem/FilePathEqualityComparer.cs

```diff
@@ -0,0 +1,63 @@
+using System.Runtime.InteropServices;
+
+namespace Mjml.Net.Includes.Filesystem;
+
+/// <summary>
+/// An equality comparer for paths in a filesystem. Works with both relative and absolute paths.
+/// </summary>
+public class FilePathEqualityComparer : IEqualityComparer<string>
+{
+    private readonly string? workingDirectory;
+    private readonly StringComparison equalityComparer;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="FilePathEqualityComparer"/> class.
+    /// </summary>
+    /// <param name="workingDirectory">A base directory to compare relative and absolute paths. Must be absolute to handle paths correctly.</param>
+    /// <param name="isPathCaseSensitive">Determine if paths should be treated as case-sensitive.
+    ///     Null value (default) means that this value will be computed based on current OS platform.</param>
+    public FilePathEqualityComparer(string? workingDirectory = null, bool? isPathCaseSensitive = null)
+    {
+        this.workingDirectory = workingDirectory;
+
+        // For different OS we have different path case-sensitivity behavior.
+        isPathCaseSensitive ??= RuntimeInformation.IsOSPlatform(OSPlatform.Linux) ||
+                                RuntimeInformation.IsOSPlatform(OSPlatform.FreeBSD);
+
+        equalityComparer = isPathCaseSensitive == true
+            ? StringComparison.InvariantCulture
+            : StringComparison.InvariantCultureIgnoreCase;
+    }
+
+    public bool Equals(string? x, string? y)
+    {
+        if (x == null && y == null)
+        {
+            return true;
+        }
+
+        if (x == null || y == null)
+        {
+            return false;
+        }
+
+        if (Path.IsPathRooted(x) != Path.IsPathRooted(y))
```

Comment:
Because it will depend on the app working directory. It is not always a valid scenario, since in most cases we combining relative paths.
Let's say we have an app that is launched from `C:/App/`
In this case, relative path `./test.mjml` will point to `C:/App/test.mjml` after resolving via `Path.GetFullPath`. But it's not always a correct folder since we can have our templates at `C:/Templates/` folder, so `C:/Templates/test.mjml` and `C:/App/test.mjml` are different file paths and path comparer must respond false to this. That's why I introduced a working directory for this comparer. Without working directory specified we cannot guarantee that this is the correct file.

But I agree that this part is not ideal, if you have any ideas here - let me know.
My first thought is to make workingDirectory parameter required and it must be absolute path. But this case is not ideal, since in most cases we operate with relative paths, so it won't be really needed and probably confusing.

Reply:
Basically, this file is not a file resolver. It's a path equality comparer. So, its job is to determine if two paths are equal or not and return boolean. 
1. It's easy task when you have two absolute paths. It's always possible to say if they are equal or not. 
2. When they both relative it's also an easy task, since app working directory will be the same for both relative paths when computing absolute path - app working directories will be common prefix for both paths, so it doesn't affect comparison result.
3. Troubles might appear when one path is absolute and another is relative. We need to have a base path for relative path in this case to have an ability to compare them.

But probably, we should not return false when base path is not specified. I think we should proceed with app working directory instead. Maybe in this case we can have some sort of warning.

In any case I think we should have such comparer in place and provide ability to write custom one.

Reply:
I think it depends on the order:

If you do it like this:

1. Get relative path.
2. Check cache with relative path and comparer
3. Resolver absolute path
4. Load file

Then you need the comparer.

But if you change the order to:

1. Get relative path.
3. Resolver absolute path
3. Check cache with absolute path
4. Load file

Then I don't see why this is needed.

The only problem is that the dictionary is public, which is more an implementation detail and used for testing, not production code. When you add method like AddToCache() instead you can make this normalization there as well.

But here is the composition topic again. I really wanted to have my stateful file loader because I hated how the include component looked like. So what I would do is to build a FileSystemFileLoader on top of that that also accepts an optional cache instance.